File: 1603-designParkingSystem/designParkingSystem.go
package designparkingsystem

const (
	BIG    = 1
	MEDIUM = 2
	SMALL  = 3
)

type Capacity struct {
	filled int
	total  int
}
type ParkingSystem struct {
	big    Capacity
	medium Capacity
	small  Capacity
}

func Constructor(big int, medium int, small int) ParkingSystem {
	return ParkingSystem{Capacity{0, big}, Capacity{0, medium}, Capacity{0, small}}
}

func (this *ParkingSystem) AddCar(carType int) bool {
	switch carType {
	case BIG:
		if this.big.filled < this.big.total {
			this.big.filled++
			return true
		}
	case MEDIUM:
		if this.medium.filled < this.medium.total {
			this.medium.filled++
			return true
		}
	case SMALL:
		if this.small.filled < this.small.total {
			this.small.filled++
			return true
		}
	}
	return false
}

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * obj := Constructor(big, medium, small);
 * param_1 := obj.AddCar(carType);
 */

File: 1603-designParkingSystem/designParkingSystem.rs
struct ParkingSystem {
    big : (i32,i32),
    medium : (i32,i32),
    small : (i32,i32),
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl ParkingSystem {

    // I couldn't make work defining BIG MEDIUM AND SMALL as constants

    fn new(big: i32, medium: i32, small: i32) -> Self {
        ParkingSystem {
            big: (0,big),
            medium: (0,medium),
            small: (0,small),
        }
    }
    
    fn add_car(&mut self, car_type: i32) -> bool {
        println!("{} {} {}", self.big.0, self.medium.0, self.small.0);
        match car_type {
            1 => {
                if self.big.0 < self.big.1 {
                    self.big.0 += 1;
                    true
                } else {
                    false
                }
            },
            2 => {
                if self.medium.0 < self.medium.1 {
                    self.medium.0 += 1;
                    true
                } else {
                    false
                }
            },
            3 => {
                if self.small.0 < self.small.1 {
                    self.small.0 += 1;
                    true
                } else {
                    false
                }
            },
            _ => false,
        }
    }
}

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * let obj = ParkingSystem::new(big, medium, small);
 * let ret_1: bool = obj.add_car(carType);
 */